Challenge for 2FA every 8 hours + Refactoring (alpha)

xoen opened the pull request:
(same as [`dev` PR](https://github.com/ministryofjustice/analytics-platform-auth0/pull/17) but different 2FA label and `DISABLED_CLIENT` for `alpha`)

Before this change we were challenging for 2FA at every login.
This was alright but as we want to have a more streamlined user experience
and enable silent SSO this is not a problem.

The rule now check when was last time a user logged in (by getting the full
Auth0 profile which contains `last_login`) and only challenge if he/she
didn't login in the last 8 hours.

### NOTE
From empiric experience it seems like the user goes through this rule twice. 
When user goes through rule the first time their `last_login` field is updated even if they didn't enter the 2FA code yet (unfortunately).

This is the reason why if user logged in less than 5 seconds I consider that as a non-login and re-enforce 2FA (otherwise they will not be challenged). 
I also tested that if user waits before clicking on "login" button they will still be challenged for 2FA code.


### Ticket
Part of https://trello.com/c/PqvTE1VQ/672-4-update-auth0-rule-to-challenge-for-2fa-every-8h

xoen committed:
Challenge for 2FA every 8 hours + Refactoring
Before this change we were challenging for 2FA at every login.
This was alright but as we want to have a more streamlined user experience
and enable silent SSO this is not a problem.

The rule now check when was last time a user logged in (by getting the full
Auth0 profile which contains `last_login`) and only challenge if he/she
didn't login in the last 8 hours.

NOTE: From empiric experience it seems like the user goes through this
      rule twice. When user goes through rule the first time their `last_login`
      field is updated even if they didn't enter the 2FA code yet (unfortunately).

      This is the reason why if user logged in less than 5 seconds I consider
      that as a non-login and re-enforce 2FA (otherwise they will not be
      challenged). I also tested that if user waits before clicking on
      "login" button they will still be challenged for 2FA code.

andyhd commented on rules/Multifactor-Google-Authenticator-Do-Not-Rename.js:
```diff
@@ -1,35 +1,74 @@
-        mfa_enabled_connection &&
-        mfa_disabled_clients
-    ) {
+    if ( !enabled_for_user || !enabled_for_connection || !enabled_for_client ) {
```
minor thing, but this might be nicer if you invert the logic and have `if (disabled_for_user || disabled_for_connection ...`

andyhd replied:
Also a minor thing, but [Do not add spaces inside parentheses](https://github.com/airbnb/javascript#whitespace--in-parens)

andyhd commented on rules/Multifactor-Google-Authenticator-Do-Not-Rename.js:
```diff
@@ -1,35 +1,74 @@
-        mfa_disabled_clients
-    ) {
+    if ( !enabled_for_user || !enabled_for_connection || !enabled_for_client ) {
+        console.log('2FA disabled for user, connection or client');
```
Do we want `console.log` in rules?

andyhd commented on rules/Multifactor-Google-Authenticator-Do-Not-Rename.js:
```diff
@@ -1,35 +1,74 @@
 
-            // optional, the label shown in the authenticator app
-            issuer: 'MOJ Analytical Platform (Alpha)',
+            return callback(null, user, context);
```
Not sure, but I think you don't need `return` here

andyhd commented on rules/Multifactor-Google-Authenticator-Do-Not-Rename.js:
```diff
@@ -1,35 +1,74 @@
+        })
+        .catch(function(err) {
+            enableMFA(context);
+            return callback(null, user, context);
```
Or here

andyhd commented on rules/Multifactor-Google-Authenticator-Do-Not-Rename.js:
```diff
@@ -1,35 +1,74 @@
-            // optional, the key to use for TOTP. By default one is generated for you
-            // key: '{YOUR_KEY_HERE}',
 
+    // Utility functions
```
Useless comment

xoen replied:
💔

andyhd commented on rules/Multifactor-Google-Authenticator-Do-Not-Rename.js:
```diff
@@ -1,35 +1,74 @@
+        if (user_profile.last_login) {
+            var secs_from_last_login = (new Date() - new Date(user_profile.last_login)) / 1000;
+
+            if (secs_from_last_login < 5) {
```
This this the fix for the problem you were having? I think a comment might be useful here (😱)

andyhd reviewed (approved):
👍 a few minor suggestions

xoen committed:
PR amendments:
- inverted 2FA exceptions logic for readability
- added note on 5 seconds threashold to consider a user logged in
- removed `console.log()`
- removed spaces in parentheses
- removed comments

xoen commented:
Thanks @andyhd - addressed in this commit: https://github.com/ministryofjustice/analytics-platform-auth0/pull/18/commits/91e16e94ad3809af94d13d6a168e690d940d940d

FYI: Merging/Releasing.
